### PR TITLE
Reject incoming DNS requests while restoring/upgrading (bsc#966419)

### PR DESCRIPTION
**Pull request by rhafer**
This is needed to avoid a running cloud accessing the (incomplete) DNS info
that the admin node would handout when the installer runs as that might cause
services to fail when upgrading the admin node.

Ideally we would reconfigure bind to only listen on 127.0.0.1 during that
stage, but that requires some more changes as dnsmasq is already running on
that (to be done in a later commit).

https://bugzilla.suse.com/show_bug.cgi?id=966419

### CODE REVIEW

File: chef/cookbooks/bind9/recipes/default.rb [hunk 1]
```diff
@@ -347,6 +347,31 @@ def make_zone(zone)
 
 # We would like to bind service only to ip address from admin network
 admin_addr = Chef::Recipe::Barclamp::Inventory.get_network_by_type(node, "admin").address
+# if we're the admin node and no other node exists yet just listen on loop back
```
Comment (vuntz) [inline chef/cookbooks/bind9/recipes/default.rb]: Update comment for restore-only

Comment (rhafer) [inline chef/cookbooks/bind9/recipes/default.rb]: Oops, that's completely bogus anyway and from different attempt to fix the problem. I'll rephrase.

**Comment by rhafer** [comment]
Fixed comment.

I also removed the check for the number of nodes, it's not needed anymore and actually had the potential of being racey.

File: chef/cookbooks/bind9/recipes/default.rb [hunk 2]
```diff
@@ -348,6 +348,36 @@ def make_zone(zone)
...
+  bash "block incoming DNS request during install/upgrade" do
+    code <<-EOH
+      iptables -I INPUT -i #{admin_if} -p udp -m udp --dport 53 \
+        -j REJECT -m comment --comment "UPGRADE_DNSBLOCK"
+      iptables -I INPUT -i #{admin_if} -p tcp -m tcp --dport 53 \
+        -j REJECT -m comment --comment "UPGRADE_DNSBLOCK"
+    EOH
+    not_if "iptables -L INPUT | grep -q UPGRADE_DNSBLOCK"
+  end
+else
```
Comment (vuntz) [inline chef/cookbooks/bind9/recipes/default.rb]: You can probably do `elsif node["crowbar"]["admin_node"]`

Comment (rhafer) [inline chef/cookbooks/bind9/recipes/default.rb]: Right. Done.

File: chef/cookbooks/bind9/recipes/default.rb [hunk 3]
```diff
@@ -348,6 +348,36 @@ def make_zone(zone)
 # We would like to bind service only to ip address from admin network
 admin_addr = Chef::Recipe::Barclamp::Inventory.get_network_by_type(node, "admin").address
 
+# When we're restoring the admin node from backup or upgrade data,
+# reject incoming DNS traffic to avoid sending wrong results to running
+# clients.
+# FIXME: A cleaner approach would be to restrict the listner address of
```
Comment (vuntz) [inline chef/cookbooks/bind9/recipes/default.rb]: s/listner/listener/ but I can live with it

**Comment by vuntz** [comment]
+1

**Comment by vuntz** [comment]
@rhafer do you think it makes sense to have a followup card to clean this up?

**Comment by rhafer** [comment]
> do you think it makes sense to have a followup card to clean this up?

Yes. I'd also leave the bug open. (With lower priority)

**Comment by vuntz** [comment]
+1

**Comment by aplanas** [comment]
+1

**Comment by rhafer** [comment]
FYI, I just did another successful test run with this patch on a complete new deployment.

**Pull request merged**